### PR TITLE
port to typescript; split usable parts; form proper cli tool; docs++

### PR DESCRIPTION
**Pull request by sdrsdr**
Let's have a tool we're not ashamed of!

### CODE REVIEW

File: http-digest-auth/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      noformat=true;
+      continue;
+    }
+    default: {
+      usage();
+    }
+  }
+}
+
+if (password == '' || host == '') {
```
Comment (Mircho) [inline http-digest-auth/src/index.ts]: In the context of having this piece of code as a tool, password can be empty. Which is the usual case.
Consider having it optional.

File: http-digest-auth/src/shellyHttpCall.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  let resp = HexHash(username + ":" + authParams.realm + ":" + password);
+  resp += ":" + authParams.nonce;
+  resp += ":1:" + authParams.cnonce + static_noise_sha256;
+
+  authParams.response = HexHash(resp);
+
+};
+
+
+export const shellyHttpUsername = "admin"; // always
```
Comment (Mircho) [inline http-digest-auth/src/shellyHttpCall.ts:65]: Do we need to export this?

File: http-digest-auth/src/shellyHttpCall.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+      if (_value != 'auth') {
+        throw new Error("WWW-Authenticate heder is requesting unusial qop:" + _value + " instead of auth");
+      }
+      continue;
+    }
+
+    authParams[_key.trim()] = _value.trim();
+  }
+  if (!isauthParams(authParams)) {
+    throw new Error("invalid WWW-Authenticate heder from device?!");
```
Comment (Mircho) [inline http-digest-auth/src/shellyHttpCall.ts]: heder->header (also above)

File: http-digest-auth/src/shellyHttpCall.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      let buffer = Buffer.alloc(0);
+
+      if (response.statusCode == 401) {
+        // Not authenticated
+        if (!try_authenticate) {
+          return reject(new Error("Failed to authenticate!"));
+        }
+        //look up the challenge header
+        let authHeader = response.headers["www-authenticate"];
+        console.error("calculating digest auth from WWW-Authenticate heder:", authHeader);
```
Comment (Mircho) [inline http-digest-auth/src/shellyHttpCall.ts]: Leftover?

File: http-digest-auth/src/shellyHttpCall.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      },
+      timeout: 500
+    };
+
+    const req = http.request(options, async (response) => {
+      let buffer = Buffer.alloc(0);
+
+      if (response.statusCode == 401) {
+        // Not authenticated
+        if (!try_authenticate) {
```
Comment (Mircho) [inline http-digest-auth/src/shellyHttpCall.ts]: We should also bail out if password is empty. Having it empty is equivalent to not having password thus we can't respond to the challenge

File: http-digest-auth/src/shellyHttpCall.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+    authParams[_key.trim()] = _value.trim();
+  }
+  if (!isauthParams(authParams)) {
+    throw new Error("invalid WWW-Authenticate heder from device?!");
+  }
+  return authParams;
+}
+
+export function shellyHttpCall(postdata: JRPCPost_t, host: string, password: string, try_authenticate: boolean = true): Promise<string> {
```
Comment (Mircho) [inline http-digest-auth/src/shellyHttpCall.ts]: Maybe replace try_authenticate with failOnAuthChallenge and reverse the logic - if true and we get 401 bail out

**Review by Mircho — COMMENTED** [review]
First round done. At the moment coding style is not very consistent. 
If we assume https://google.github.io/styleguide/jsguide.html as a guide there is a lot to be done.
One step at a time.

File: http-digest-auth/src/shellyHttpCall.ts (hunk 2 above)
Comment (sdrsdr) [reply http-digest-auth/src/shellyHttpCall.ts:65]: A look at this exports like a documentation. If you want to do other things with that magic string why not take it from the source. Latter if we change things and change the string you get it changed automatically plus it's just a ref that is exposed the real string inside the library can't be changed ...

File: http-digest-auth/src/shellyHttpCall.ts (hunk 4 above)
Comment (sdrsdr) [reply http-digest-auth/src/shellyHttpCall.ts]: I really like to indicate that the fetch was done via authenticated session. I agree a library what shellyHttpCall.ts actually is should not print but I can't think of quick elegant solution so I kinda left it as is opting for the trick with -q option. Suggestions are welcome

Comment (sdrsdr) [reply http-digest-auth/src/shellyHttpCall.ts]: Ok the indication can be done based on presence of .auth in postData ..

**Pull request merged**